### ISSUE TITLE
Type not working with custom rule

### ISSUE THREAD

**Issue by Dylan-Prins** [Azure/PSRule.Rules.Azure#2278]
Hi,

I want to create a custom rule that that checks parameter descriptions for values. I have 2 problems:

- type not working: -Type 'Microsoft.Resources/deployments'
Rule is not being executed
- If I use the commented part then the rule errors

```powershell
Rule 'Azure.Template.Parameters.RequiredOrOptional' -Type 'Microsoft.Resources/deployments' {
    Recommend 'Parameters should have the correct description'

    $parameters = @($TargetObject.properties.template.parameters.PSObject.properties)
    if ($parameters.Length -eq 0) {
        return $Assert.Pass()
    }
    foreach ($Parameter in $Parameters) {
        $Assert.StartsWith($Parameter.Value, 'metadata.description', @('Optional', 'Required')).
        Reason('The description of the optional parameter "{0}" doesn''t start with "Optional"', $Parameter.Name)

        # if (($Parameter.Value | Get-Member).Name -contains 'defaultValue') {
        #     $Assert.StartsWith($Parameter.Value, 'metadata.description', 'Optional').
        #     Reason('The description of the optional parameter "{0}" doesn''t start with "Optional"', $Parameter.Name)
        # }

        # if (($Parameter.Value | Get-Member).Name -notcontains 'defaultValue') {
        #     $Assert.StartsWith($Parameter.Value, 'metadata.description', 'Required').
        #     Reason('The description of the optional parameter "{0}" doesn''t start with "Required"', $Parameter.Name)
        # }
        # $Assert.Fail("Failed tests")
    }
}
```

**Comment by BernieWhite** [comment]
@Dylan-Prins When configuring custom rules you need to configure binding options for to use `-Type 'Microsoft.Resources/deployments'`. See https://azure.github.io/PSRule.Rules.Azure/customization/enforce-custom-tags/#binding-type

For example:

```yaml
# Configure binding options
binding:
  targetType:
  - 'resourceType'
  - 'type'
```

For the other issue, calling `$Assert.*` won't drop out of the loop so `$Assert.Fail("Failed tests")` would always be processed (unless there is no parameters). If any conditions fail, the whole rule fails. Removing that line, is required.

However you could also update as follows:

```powershell
Rule 'Azure.Template.Parameters.RequiredOrOptional' -Type 'Microsoft.Resources/deployments' {
    Recommend 'Parameters should have the correct description'

    $parameters = @($TargetObject.properties.template.parameters.PSObject.properties)
    if ($parameters.Length -eq 0) {
        return $Assert.Pass()
    }
    foreach ($Parameter in $Parameters) {
        $parameterPrefix = 'Required'
        if ($Assert.HasField($Parameter, 'defaultValue').Result) {
            'Optional'
        }
        $Assert.StartsWith($Parameter, 'metadata.description', $parameterPrefix).
            Reason(
                'The description of the {1} parameter "{0}" doesn''t start with "{2}"',
                $Parameter.Name,
                $parameterPrefix.ToLower(),
                $parameterPrefix
            )
    }
}
```

---

I hope that helps.

**Comment by Dylan-Prins** [comment]
Perfect! this helps me a lot.

But I still have problems with the right context. I See it running the same context more then once

Wrong context:
```text
Template: modules/azure/features/WebAvailabilityTest/examples/minimal/example.bicep:0:0

  Recommend:
  Parameters should have the correct description

  Reason:
  - The description of the required parameter "" doesn't start with "Required"
 ```
Right context:
```text
Template: modules/azure/features/WebAvailabilityTest/examples/minimal/example.bicep:0:0

  Recommend:
  Parameters should have the correct description

  Reason:
  - The description of the required parameter "tags" doesn't start with "Required"
```

I tried the `-If { (IsParentDeployment) }`, but that sets it on the wrong context.

Working rule:
```powershell
Rule 'Azure.Template.Parameters.RequiredOrOptional' -Type 'Microsoft.Resources/deployments' {
    Recommend 'Parameters should have the correct description'

    $parameters = @($TargetObject.properties.template.parameters.PSObject.properties)
    if ($parameters.Length -eq 0) {
        return $Assert.Pass()
    }
    foreach ($Parameter in $Parameters) {
        $parameterPrefix = 'Required'
        if ($Assert.HasField($Parameter, 'value.defaultValue').Result) {
            $parameterPrefix = 'Optional'
        }
        $Assert.StartsWith($Parameter, 'value.metadata.description', $parameterPrefix).
        Reason(
            'The description of the {1} parameter "{0}" doesn''t start with "{2}"',
            $Parameter.Name,
            $parameterPrefix.ToLower(),
            $parameterPrefix
        )
    }
}
```

**Comment by BernieWhite** [comment]
@Dylan-Prins No worries.

When PSRule gets a Bicep or ARM template it will report the parent file as one deployment, then any nested deployments/ modules it will report as subsequent deployments.

In Bicep it's common to call modules and modules in modules so you might have a few deployments to unpack. The names of the deployments should give it away.

Based on your question I'm assuming, you want the nested deployments. Is that correct?

You're 100% on the right track. But maybe try `-If { !(IsParentDeployment) }`.

---

Side note: you can use `Write-Debug` and `Write-Verbose` in your code to debug further, locally run something like `Assert-PSRule -Module "PSRule.Rules.Azure" -Path "./.ps-rule/" -Format File -InputPath '.' -Outcome Fail,Error -Debug -Verbose`
